Add Filtering of Segmentation Results by LULC Mask for Agricultural Fields

mishagrol opened the pull request:
**Motivation**

In many regions, abandoned fields and other land plots are often misidentified as agricultural fields by segmentation models. These inaccuracies can negatively impact data analysis and decision-making. To reduce such errors, we propose an additional filtering step using a Land Use/Land Cover (LULC) mask. This approach ensures a more precise identification of agricultural fields and enhances the quality of data analysis.

**Implemented Pipeline**

The new pipeline introduces the following improvements:
- Download the LULC map from Planetary STAC using the bounding box of polygonized fields.
- Vectorize the LULC map to extract polygons for the agriculture class.
- Filter polygonized fields based on a minimum area threshold (m²).
- Perform a spatial join between fields and LULC polygons.
- Retain only fields that intersect with agricultural LULC polygons.

This additional filtering step significantly improves the accuracy of field segmentation.
Examples

The following examples demonstrate the pipeline for two regions: *Austria* and a small part of *Russia (Kaliningrad region)*.

**Example: Austria**
```bash
ftw inference download \
 --win_a S2B_MSIL2A_20210617T100559_R022_T33UUP_20210624T063729 \
 --win_b S2B_MSIL2A_20210925T101019_R022_T33UUP_20210926T121923 \
 -f -o austria.tif --bbox=12.7,48.4,12.97,48.57

ftw inference run austria.tif -f -o austria-inf.tif  -m 3_Class_FULL_FTW_Pretrained.ckpt

ftw inference polygonize -f --simplify 15 -o austria-inf.geojson austria-inf.tif

ftw inference filter_by_lulc -f austria-inf.geojson -o austria-inf-lulc.geojson \
 --minimal_area_m2 1000 --lulc_path LULC.tif --lulc_year 2023
```

**Example: Russia (Kaliningrad region)**

```bash
ftw inference download \
 --win_a S2A_MSIL2A_20241018T095031_R079_T34UEF_20241018T124947 \
 --win_b S2B_MSIL2A_20240722T093549_R036_T34UEF_20240722T142653 \
 -f -o kaliningrad.tif --bbox=21.75,54.542,22.400,54.739

ftw inference run kaliningrad.tif -f --mps_mode \
 -o kaliningrad-inf.tif -m 3_Class_FULL_FTW_Pretrained.ckpt    

ftw inference polygonize -f --simplify 15 -o kaliningrad-inf.geojson kaliningrad-inf.tif

ftw inference filter_by_lulc -f kaliningrad-inf.geojson -o kaliningrad-inf-lulc.geojson \
 --minimal_area_m2 1000 --lulc_path LULC.tif --lulc_year 2023
```

**Maps for the Russian Fields Example**

<img width="1081" alt="Снимок экрана 2025-03-20 в 15 19 59" src="https://github.com/user-attachments/assets/0cf332eb-ab71-4957-99c1-6899806cb0da" />


**Key Benefits of This Update**

✅ More accurate field segmentation by removing non-agricultural areas.
✅ Automated integration of LULC data for filtering.
✅ Flexible approach that allows adjustments in LULC class, area threshold, and region selection.
✅ Improved reliability of inference outputs for real-world agricultural applications.

This enhancement significantly improves the segmentation results, making them more useful for downstream analysis. 🚀

hannah-rae commented:
Yay!!! I have a couple of questions: 
1. When you say it "significantly improves the segmentation results", is this assessed qualitatively/visually or did you recompute the test set metrics after filtering?
2. Did you consider other methods of comparing the field polygons to the raster values before deciding on this approach? For example I'm wondering if you thought about other rules for the spatial join than intersection (e.g. field within an LULC ag polygon) or doing some math on the raster values within field polygons (e.g. keep if the mode is agriculture). I think your approach is reasonable but curious if other paths were considered.
3. It looks like you chose the ESRI/Impact Observatory LULC map but there are other options like [WorldCover](https://planetarycomputer.microsoft.com/dataset/esa-worldcover) available on Planetary Computer. Map performance varies globally and users may want to choose a different map for their region. What about making which map to use an argument (or maybe do this in a future enhancement)?

mishagrol commented:
@hannah-rae thank you for your questions and comments! 

I improved some parts of code and added selection of LULC source. Detailed answers below:


!. First, I assessed the improvement qualitatively by visually comparing the segmentation results. I focused on how many field polygons were assigned to non-agricultural classes. Most of the incorrect classifications were forested areas or overgrown fields that haven't been cultivated for several years.

2. You're right, polygon-based operations weren't the most optimal approach. I’ve revised the implementation, and the filtering is now performed directly on the raster data. The updated version includes two functions: one retrieves the LULC raster using a bounding box, and the other compares it with the inference result. Based on the raster comparison, pixels outside agricultural areas are set to no data.

3. Thank you for your suggestion, I explored the available LULC datasets on the Planetary Computer and added the ability to choose the data source as an argument. Currently, the code supports two options:

    "io-lulc-annual-v02" ([link](https://planetarycomputer.microsoft.com/dataset/io-lulc-annual-v02))

    "esa-worldcover" ([link](https://planetarycomputer.microsoft.com/dataset/esa-worldcover))
    This makes the implementation more flexible and region-adaptable.

hannah-rae commented:
Thanks for these changes @mishagrol! Excited to have this as part of the pipeline.

cholmes commented:
Just got a chance to dig into this, and the flow and options look great. Love the ability to use different options.

My local repo is a bit borked, so I want to get that set up and try it out myself before merging in, but will try to do that soon.

Does it work with all the output options from polygonize? And can you do all the same output options, including fiboa? I think it could be good to default to parquet / fiboa. Would probably be good to just share the format writing code between the two functions (maybe you're doing that, don't have the time to fully figure out the code at the moment).

Thanks a ton for the contribution, this is a great idea and well executed.

mishagrol commented:
Hi @cholmes!
Thank you for the positive feedback and for taking the time to review my MR.

I want to clarify that the current implementation of LULC filtering occurs at the raster processing stage, not at the polygonize stage, as it was in the first version (thanks to @hannah-rae for this  recommendation). This allowed us to retain all formats, such as parquet / fiboa, and not affect the polygonize stage.

Yes, the implementation supports all output options from polygonize, including fiboa.

Below example of usage : 

```
# !/bin/bash
ftw inference download \
--win_a S2B_MSIL2A_20210617T100559_R022_T33UUP_20210624T063729 \
--win_b S2B_MSIL2A_20210925T101019_R022_T33UUP_20210926T121923 \
-f -o austria.tif --bbox=12.7,48.4,12.97,48.57

echo "Downloaded images"

ftw inference run austria.tif -f -o austria-inf.tif \
 -m 3_Class_FULL_FTW_Pretrained.ckpt

echo "Run inference"

ftw inference filter_by_lulc -f austria-inf.tif -o austria-inf-lulc.tif  --collection_name esa-worldcover

echo "Polygonize"

ftw inference polygonize austria-inf-lulc.tif

echo "Done"

```

mishagrol commented:
> It's missing a changelog entry, otherwise it's ready to be merged.

@m-mohr  thank you for comments! Should I add info about new functions in CHANGELOG.md (https://github.com/mishagrol/ftw-baselines/blob/main/CHANGELOG.md)?

cholmes commented on README.md:
```diff
@@ -246,6 +246,29 @@ Options:
+
+Usage: ftw inference filter_by_lulc [OPTIONS] INPUT
+
+  Filter the output field in GeoJSON format by field size and LULC mask.
```
Output is now GeoTIFF, yeah?

cholmes commented on README.md:
```diff
@@ -246,6 +246,29 @@ Options:
+Options:
+  -o, --out TEXT           Output filename for the filtered polygon data. If
+                           not given, defaults to the input filename with
+                           parquet extension. Available formats: 
```
And this should change with it being tiff too.

cholmes commented on README.md:
```diff
@@ -246,6 +246,29 @@ Options:
+                           not given, defaults to the input filename with
+                           parquet extension. Available formats: 
+                           .geojson (GeoJSON).
+  --minimal_area_m2 FLOAT  Minimum area size in square meters to include in
```
Is this still an option with the doing the tiff processing now?

cholmes commented:
> About README examples, in last commit I have added info to README.md (https://github.com/fieldsoftheworld/ftw-baselines/pull/75/files), should I add more examples to other files?

Was that a recent commit? If so then it maybe didn't get pushed? If not, then the readme stuff that is in there seems a bit out-dated, relative to the updates to run on the tiff and do tiff output. So I think just re-running the cli with the --help option and putting the updated thing in there is mostly sufficient (I'm not at my computer that I got the testing of this branch on at the moment, so can't try it). 

And then past that I think it's just adding `ftw inference filter_by_lulc -f austria-inf.tif -o austria-inf-lulc.tif  --collection_name esa-worldcover` to the readme, with a bit of explanation on what it does, and explain that it's an optional step you can do before polygonization. Like somewhere after https://github.com/fieldsoftheworld/ftw-baselines/blob/main/README.md#sample-prediction-output-austria-patch-red---fields

cholmes commented:
Sorry for the slow review! I've been slammed. Improvements look great. I'll try to find some time to test it out and then merge in, but I'm out on vacation soon, so it may be a bit. If anyone else is able to run it and confirm it works I'm happy to merge.

hannah-rae commented:
Closed by #130
